### PR TITLE
Cron submit tests

### PR DESCRIPTION
**Pull request by cryptoAtwill**
## Changes

- Adding unit tests to `set_membership` and `submit_cron`.
- Found an edge case where the cron submission execution might be stuck, see `execute_next_cron_epoch` for more details.

## Tests

```
cargo test --all
```

### CODE REVIEW

File: gateway/src/lib.rs [hunk 1]
```diff
@@ -880,6 +881,74 @@ impl Actor {
...
+                        msg: m,
+                        wrapped: false,
+                    },
+                )?;
+            }
+        }
+        Ok(RawBytes::default())
+    }
+
+    /// Externally trigger cron submission epoch. This is an edge case to ensure none of the epoches
```
Comment (adlrocha) [inline gateway/src/lib.rs:918]: I don't think we need this external function, we can include a lazy check with every cron checkpoint submission to see if there is something to be executed. We could even check when something is executed if any subsequent epoch should also be executed. The issue I see with the latter approach is that if you have more than execution pending this may run our of gas, that is why I think is better to run just one execution per submission. 

The logic could be integrated in submit_cron and work as follows:
- The submission is performed. 
- If the submission triggers an execute we do so and it is done. 
- If not, we check the last executed and the next epoch and see if it can be executed. This lazy check will ensure that no checkpoint is ever stuck.

**Review by adlrocha — CHANGES_REQUESTED** [review]
The test are great for an initial coverage. My only concern is the external execution method, I don't think we need it as mentioned in the comments. Let me know if it is clear or you want to discuss sync/

File: gateway/src/lib.rs [hunk 2]
```diff
@@ -820,146 +821,21 @@ impl Actor {
...
-    fn submit_cron(rt: &mut impl Runtime, params: CronCheckpoint) -> Result<RawBytes, ActorError> {
+    fn submit_cron(
+        rt: &mut impl Runtime,
+        checkpoint: CronCheckpoint,
+    ) -> Result<RawBytes, ActorError> {
         // submit cron can only be performed by signable addresses
         rt.validate_immediate_caller_type(CALLER_TYPES_SIGNABLE.iter())?;
 
-        let params_epoch = params.epoch;
+        Self::execute_next_cron_epoch(rt)?;
```
Comment (adlrocha) [inline gateway/src/lib.rs:831]: Take this down and perform only one execution per submission. Add a note of why we do this.

File: gateway/tests/gateway_test.rs [hunk 3]
```diff
@@ -1143,3 +1147,312 @@ fn test_apply_msg_match_target_subnet() {
...
+            submission
+                .has_submitted(rt.store(), &Address::new_id(i))
+                .unwrap(),
+            false
+        );
+    }
+}
+
+#[test]
+fn test_submit_cron_sequential_execution() {
```
Comment (adlrocha) [inline gateway/tests/gateway_test.rs:1382]: Add a few tests to see that the message is actually sent.

Comment (cryptoAtwill) [reply gateway/tests/gateway_test.rs:1382]: added

**Review by adlrocha — APPROVED** [review]
Amzing job, I think we can merge. Quick question, once this is merged, would you rather have a final pass to the `cron_submit` branch and merge it to `main`, or do you prefer to get the bottom-up refactor in in this batch?

**Comment by cryptoAtwill** [comment]
@adlrocha I think working on `cron_impl` first. If merged to main, likely downstream needs to be updated. There could be changes in bottom up, we can update golang and `ipc-client` one shot.
